src/simulator/halo_simulator.py

```diff
@@ -0,0 +1,217 @@
+        Returns:
+            A ReachPoint representing the differentially private estiamte of
+            the reach that would have been obtained for this spend allocation.
+            This estimate is obtained by simulating the construction of
+            Liquid Legions sketches, one per publisher, combining them, and
+            adding differentially private noise to the result.
+        """
+        combined_sketch = self._publishers[0].liquid_legions_sketch(spends[0])
+        estimator = StandardizedHistogramEstimator(
+            max_freq=max_frequency, epsilon=budget.epsilon, epsilon_split=epsilon_split
+        )
+        for i in range(1, len(spends)):
+            sketch = self._publishers[i].liquid_legions_sketch(spends[i])
+            combined_sketch = StandardizedHistogramEstimator.merge_two_sketches(
+                combined_sketch, sketch
+            )
+        frequencies = [
+            round(x) for x in estimator.estimate_cardinality(combined_sketch)
+        ]
+
+        # TODO(jiayu,pasin): Does this look right?
+        self._privacy_tracker.append(
+            NoisingEvent(
+                budget,
+                DP_NOISE_MECHANISM_GAUSSIAN,
+                {"epsilon": budget.epsilon, "epsilon_split": epsilon_split},
+            )
+        )
+
+        # convert result to a ReachPoint
+        impressions = self._data_set.impressions_by_spend(spends)
+        kplus_reaches = ReachPoint.frequencies_to_kplus_reaches(frequencies)
+        return ReachPoint(impressions, kplus_reaches, spends)
+
+    def simulated_venn_diagram_reach_by_spend(
+        self,
+        spends: List[float],
+        budget: PrivacyBudget,
+        max_frequency: int = 1,
+    ) -> List[Tuple[Set, ReachPoint]]:
+        """Returns a simulated differentially private Venn diagram reach estimate.
+
+        For each subset of publishers, computes a differentially private
+        reach and frequency estimate for those users who are reached by
+        all and only the publishers in that subset.
+
+        Args:
+            spends:  The hypothetical spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i.
+            budget:  The amount of privacy budget that can be consumed while
+              satisfying the request.
+            max_frequency:  The maximum frequency for which to report reach.
+        Returns:
+            A list of pairs (S, R), where S specifies a subset of publishers
+            and R is a ReachPoint representing the differentially private
+            estimate of the number of people reached in this subset.
+            The set S is given as a subset of the integers 0..p-1, where p
+            is the number of publishers.
+        """
+        raise NotImplementedError()
+
+    def simulated_reach_curve(
+        self, publisher_index: int, budget: PrivacyBudget
```

Comment:
nit: formatting seems different than the code above. You had each param on another line on other functions.

Reply:
All of the files have been formatted with the black formatter.  (See the README for more information.)  I tried changing the signature so that it listed one parameter per line, but black reverted it.  So, I think we have to live with this.